aws-datastore/src/main/java/com/amplifyframework/datastore/appsync/AppSyncClient.java

```diff
@@ -118,10 +119,16 @@ public static AppSyncClient via(@NonNull GraphQLBehavior api,
             }
         };
         final Consumer<ApiException> failureConsumer =
-            failure -> onFailure.accept(new DataStoreException(
-                "Failure performing sync query to AppSync.",
-                failure, AmplifyException.TODO_RECOVERY_SUGGESTION
-            ));
+            failure -> {
+            if(failure instanceof ApiException.ApiIrrecoverableException){
+                onFailure.accept(new DataStoreException.IrRecoverableException(
+                        "Failure performing sync query to AppSync.", AmplifyException.TODO_RECOVERY_SUGGESTION));
+            } else {
+                onFailure.accept(new DataStoreException(
+                        "Failure performing sync query to AppSync.",
+                        failure, AmplifyException.TODO_RECOVERY_SUGGESTION
+                ));
+            } };
```

Comment:
I don't think you need a `DataStoreException.IrRecoverableException` class.  Instead, don't make any changes here - just wrap the `ApiException` with a `DataStoreException` like it was.  Then, in the `RetryHandler` use `ErrorInspector.contains(error, ApiException.ApiIrrecoverableException.class)` to determine if it should be retried or not.    The `ErrorInspector` utility traverses the nested causes and checks each one to see if it matches the type you are looking for.